server/src/main/java/org/elasticsearch/transport/InboundDecoder.java

```diff
@@ -84,33 +92,39 @@ public int internalDecode(ReleasableBytesReference reference, Consumer<Object> f
                     return 0;
                 } else {
                     this.decompressor = decompressor;
-                    fragmentConsumer.accept(this.decompressor.getScheme());
+                    fragmentConsumer.accept(channel, this.decompressor.getScheme());
                 }
             }
             int remainingToConsume = totalNetworkSize - bytesConsumed;
             int maxBytesToConsume = Math.min(reference.length(), remainingToConsume);
-            ReleasableBytesReference retainedContent;
-            if (maxBytesToConsume == remainingToConsume) {
-                retainedContent = reference.retainedSlice(0, maxBytesToConsume);
-            } else {
-                retainedContent = reference.retain();
-            }
 
             int bytesConsumedThisDecode = 0;
             if (decompressor != null) {
-                bytesConsumedThisDecode += decompress(retainedContent);
+                bytesConsumedThisDecode += decompress(
+                    maxBytesToConsume == remainingToConsume ? reference.slice(0, maxBytesToConsume) : reference
+                );
                 bytesConsumed += bytesConsumedThisDecode;
                 ReleasableBytesReference decompressed;
                 while ((decompressed = decompressor.pollDecompressedPage(isDone())) != null) {
```

Comment:
I wonder if a push-style interaction out of `decompress` would be easier, since then the ref-counting is mostly not an issue.

Reply:
True, I guess that would be another possible simplification and would make decompress work more similarly to how the rest of the pipeline works with this change as well. Not sure we'd want to blow this one up with making that change here also. I'd rather push that down to a follow-up to keep this somewhat tricky change small? :)